File: apps/native-bridge/scripts/doctor.py
#!/usr/bin/env python3
"""Check the local Node-to-GridDB Python/Java runtime without connecting."""

from __future__ import annotations

import platform
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
LIB_DIR = ROOT / "lib"


def main() -> int:
    import jpype
    import pyarrow

    jars = sorted(LIB_DIR.glob("*.jar"))
    required = ("gridstore.jar", "gridstore-arrow.jar")
    missing = [name for name in required if not (LIB_DIR / name).is_file()]
    if missing:
        print(f"Missing public JAR(s): {', '.join(missing)}", file=sys.stderr)
        return 1

    if not jpype.isJVMStarted():
        jpype.startJVM(classpath=[str(path) for path in jars])

    import griddb_python

    print(f"macOS: {platform.mac_ver()[0]} ({platform.machine()})")
    print(f"Python: {platform.python_version()}")
    print(f"JPype: {jpype.__version__}")
    print(f"PyArrow: {pyarrow.__version__}")
    print(f"JVM: {jpype.getJVMVersion()}")
    print(f"GridDB Python: {getattr(griddb_python, '__version__', '5.9.0')}")
    print(f"Public JARs: OK ({len(jars)} found)")

    advanced = LIB_DIR / "gridstore-advanced.jar"
    if not advanced.is_file():
        print(
            "Cloud JAR: MISSING (copy gridstore-advanced.jar into lib/)",
            file=sys.stderr,
        )
        return 2

    print("Cloud JAR: OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: apps/scripts/doctor.mjs
import { existsSync } from 'node:fs';
import { delimiter, join } from 'node:path';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import { commandExists } from './process-utils.mjs';

const appsRoot = fileURLToPath(new URL('..', import.meta.url));
const bridgeRoot = join(appsRoot, 'native-bridge');
const localPython = process.platform === 'win32'
  ? join(bridgeRoot, '.venv', 'Scripts', 'python.exe')
  : join(bridgeRoot, '.venv', 'bin', 'python');
const macJavaHomes = [
  '/opt/homebrew/opt/openjdk@21/libexec/openjdk.jdk/Contents/Home',
  '/usr/local/opt/openjdk@21/libexec/openjdk.jdk/Contents/Home',
];
const detectedJavaHome = process.env.JAVA_HOME
  ?? macJavaHomes.find((home) => existsSync(join(home, 'bin', 'java')));
const javaCommand = detectedJavaHome
  ? join(detectedJavaHome, 'bin', 'java')
  : 'java';

const pythonCandidates = [process.env.PYTHON_BIN, localPython, 'python3', 'python']
  .filter(Boolean);
const python = pythonCandidates.find((candidate) => commandExists(candidate));

const checks = [
  ['Node.js 18+', Number(process.versions.node.split('.')[0]) >= 18],
  ['npm', commandExists(process.platform === 'win32' ? 'npm.cmd' : 'npm')],
  ['Python', Boolean(python)],
  ['Java', commandExists(javaCommand)],
  ['Maven', commandExists(process.platform === 'win32' ? 'mvn.cmd' : 'mvn')],
  ['GridDB public JAR', existsSync(join(bridgeRoot, 'lib', 'gridstore.jar'))],
  ['GridDB Arrow JAR', existsSync(join(bridgeRoot, 'lib', 'gridstore-arrow.jar'))],
  ['GridDB Cloud JAR', existsSync(join(bridgeRoot, 'lib', 'gridstore-advanced.jar'))],
  ['Environment file', existsSync(join(bridgeRoot, '.env'))],
];

if (python) {
  const pythonEnvironment = { ...process.env };
  if (detectedJavaHome) {
    pythonEnvironment.JAVA_HOME = detectedJavaHome;
    pythonEnvironment.PATH = `${join(detectedJavaHome, 'bin')}${delimiter}${pythonEnvironment.PATH ?? ''}`;
  }
  if (!pythonEnvironment.JAVA_TOOL_OPTIONS?.includes('java.base/java.nio')) {
    pythonEnvironment.JAVA_TOOL_OPTIONS = [
      pythonEnvironment.JAVA_TOOL_OPTIONS,
      '--add-opens=java.base/java.nio=ALL-UNNAMED',
    ].filter(Boolean).join(' ');
  }
  const imports = spawnSync(
    python,
    [
      '-c',
      'from pathlib import Path; import jpype; jpype.startJVM(classpath=[str(path) for path in Path("lib").glob("*.jar")]); import griddb_python, pyarrow',
    ],
    { cwd: bridgeRoot, env: pythonEnvironment, stdio: 'ignore' },
  );
  checks.push(['Python GridDB runtime', imports.status === 0]);
}

const labelWidth = Math.max(...checks.map(([label]) => label.length));
for (const [label, passed] of checks) {
  console.log(`${passed ? 'PASS' : 'MISS'}  ${label.padEnd(labelWidth)}`);
}

const missing = checks.filter(([, passed]) => !passed);
if (missing.length) {
  console.error(`\n${missing.length} prerequisite(s) are missing.`);
  console.error('See README.md and run the setup command for your platform.');
  process.exitCode = 1;
} else {
  console.log('\nThe local GridDB runtime is ready.');
}
